components/org.wso2.carbon.identity.user.account.association/src/main/java/org/wso2/carbon/identity/user/account/association/handler/grant/AccountSwitchGrantHandler.java

```diff
@@ -99,9 +119,19 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
 
             return false;
         }
+        AuthenticatedUser authenticatedUser = OAuth2Util.getUserFromUserName(associatedUser.toFullQualifiedUsername());
+        if (isSubOrganizationUser(accessTokenDO)) {
+            authenticatedUser.setAccessingOrganization(accessTokenDO.getAuthzUser().getAccessingOrganization());
+            authenticatedUser.setUserResidentOrganization(accessTokenDO.getAuthzUser().getUserResidentOrganization());
+            authenticatedUser.setFederatedUser(true);
+            authenticatedUser.setFederatedIdPName(accessTokenDO.getAuthzUser().getFederatedIdPName());
+            authenticatedUser.setTenantDomain(accessTokenDO.getAuthzUser().getTenantDomain());
+            String userId = getUserId(username, tenantDomain);
+            authenticatedUser.setUserId(userId);
+            authenticatedUser.setUserName(userId + "@" + accessTokenDO.getAuthzUser().getUserResidentOrganization());
```

Comment:
Better if we can follow this approach https://github.com/wso2-extensions/identity-oauth2-grant-organization-switch/blob/main/components/org.wso2.carbon.identity.oauth2.grant.organizationswitch/src/main/java/org/wso2/carbon/identity/oauth2/grant/organizationswitch/OrganizationSwitchGrant.java#L102 and then set the required information like accessing organization and resident organization.
Trying to set the user-id may not be required.

Reply:
yes if not changed the user id handling for federated user fails.

Reply:
Any comment for the first suggession.

Reply:
Fixed